apps/onboarding-functions/src/main/java/it/pagopa/selfcare/onboarding/service/CompletionServiceDefault.java

```diff
@@ -376,4 +391,105 @@ private List<String> getDestinationMails(Onboarding onboarding) {
+            return Collections.emptyList();
+        }
+
+        return activeManagers.stream()
+                .map(UserInstitutionResponse::getUserId)
+                .toList();
+    }
+
+    private String retrieveTaxCode(String uid) {
+        return userRegistryApi.findByIdUsingGET(USERS_FIELD_LIST, uid).getFiscalCode();
+    }
+
+    private boolean isActiveManagerOnRegistries(String userTaxCode, String institutionTaxCode, Origin origin) {
+        return switch (origin) {
+            case INFOCAMERE -> isActiveManagerOnInfocamereRegistry(userTaxCode, institutionTaxCode);
+            case ADE -> isActiveManagerOnAdeRegistry(userTaxCode, institutionTaxCode);
+            default -> throw new GenericOnboardingException("Origin not supported");
+        };
+    }
+
+    private boolean isActiveManagerOnInfocamereRegistry(String userTaxCode, String institutionTaxCode) {
+        BusinessesResource businessesResource = infocamereApi.institutionsByLegalTaxIdUsingPOST(toGetInstitutionByLegalDto(userTaxCode));
+        if(Objects.isNull(businessesResource) || CollectionUtils.isEmpty(businessesResource.getBusinesses())) {
+            return false;
+        }
+
+        return businessesResource.getBusinesses().stream()
+                .anyMatch(business -> institutionTaxCode.equals(business.getBusinessTaxId()));
+    }
+
+    private GetInstitutionsByLegalDto toGetInstitutionByLegalDto(String userTaxCode) {
+        GetInstitutionsByLegalDto getInstitutionsByLegalDto = new GetInstitutionsByLegalDto();
+        GetInstitutionsByLegalFilterDto getInstitutionsByLegalFilterDto = new GetInstitutionsByLegalFilterDto();
+        getInstitutionsByLegalFilterDto.setLegalTaxId(userTaxCode);
+        getInstitutionsByLegalDto.setFilter(getInstitutionsByLegalFilterDto);
+        return getInstitutionsByLegalDto;
+    }
+
+    private boolean isActiveManagerOnAdeRegistry(String userTaxCode, String institutionTaxCode) {
+        try {
+            LegalVerificationResult legalVerificationResult = nationalRegistriesApi.verifyLegalUsingGET(userTaxCode, institutionTaxCode);
+            return legalVerificationResult.getVerificationResult();
+        } catch (WebApplicationException e) {
+            // 400 status code means that the user is not a manager of the institution
+            if (e.getResponse().getStatus() == 400) {
+                return false;
+            }
+            throw new GenericOnboardingException(String.format("Error during verify legal %s", e.getMessage()));
+        }
+    }
+
+    private void deleteManagerFromProduct(String uid, String institutionId, String productId) {
+        try (Response response = userApi.usersUserIdInstitutionsInstitutionIdProductsProductIdDelete(institutionId, productId, uid)) {
+            if (!SUCCESSFUL.equals(response.getStatusInfo().getFamily())) {
+                throw new GenericOnboardingException("Impossible to delete user: " + uid);
```

Comment:
you could providing more context can aid in debugging and maintenance....for ex: String.format("Failed to delete user '%s' from product '%s' in institution '%s'", uid, productId, institutionId));